File: scheduled-jobs-stop-after-uptime/python/check_scheduler_heartbeat.py
"""Detect a stalled Medusa v2 scheduler caused by a hung workflow step
occupying the only BullMQ worker slot (jobWorkerOptions.concurrency=1
by default on @medusajs/medusa/workflow-engine-redis). There is no
Admin API route that can kill a stuck job or restart the scheduler,
so this only flags the stall and alerts an operator to restart the
worker process. DRY_RUN=true only logs locally; DRY_RUN=false also
calls the alert webhook. Never writes anything to Medusa itself.
"""
import os
import logging
from datetime import datetime, timedelta, timezone

import requests

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("check_scheduler_heartbeat")

BASE_URL = os.environ.get("MEDUSA_BACKEND_URL", "http://localhost:9000")
EMAIL = os.environ.get("MEDUSA_ADMIN_EMAIL", "admin@example.com")
PASSWORD = os.environ.get("MEDUSA_ADMIN_PASSWORD", "supersecret")
STOCK_LOCATION_ID = os.environ.get("HEARTBEAT_STOCK_LOCATION_ID", "sloc_heartbeat")
HEARTBEAT_CRON = os.environ.get("HEARTBEAT_CRON", "*/5 * * * *")
TOLERANCE_MULTIPLIER = float(os.environ.get("TOLERANCE_MULTIPLIER", "3"))
ALERT_WEBHOOK_URL = os.environ.get("ALERT_WEBHOOK_URL")
DRY_RUN = os.environ.get("DRY_RUN", "true").lower() == "true"


def get_token():
    r = requests.post(
        f"{BASE_URL}/auth/user/emailpass",
        json={"email": EMAIL, "password": PASSWORD},
        timeout=30,
    )
    r.raise_for_status()
    return r.json()["token"]


def get_heartbeat_last_run_at(token, stock_location_id):
    headers = {"Authorization": f"Bearer {token}"}
    r = requests.get(
        f"{BASE_URL}/admin/stock-locations",
        params={"fields": "id,metadata", "limit": 100},
        headers=headers,
        timeout=30,
    )
    r.raise_for_status()
    locations = r.json()["stock_locations"]
    for loc in locations:
        if loc["id"] == stock_location_id:
            return (loc.get("metadata") or {}).get("heartbeat_last_run_at")
    return None


def _parse_field(field, lo, hi):
    """Parse one cron field (*, N, N-M, N,M, */S) into a set of allowed ints."""
    if field == "*":
        return set(range(lo, hi + 1))
    values = set()
    for part in field.split(","):
        if part.startswith("*/"):
            values.update(range(lo, hi + 1, int(part[2:])))
        elif "-" in part:
            a, b = part.split("-")
            values.update(range(int(a), int(b) + 1))
        else:
            values.add(int(part))
    return values


def _parse_cron(cron_expression):
    minute, hour, dom, month, dow = cron_expression.strip().split()
    return {
        "minute": _parse_field(minute, 0, 59),
        "hour": _parse_field(hour, 0, 23),
        "dom": _parse_field(dom, 1, 31),
        "month": _parse_field(month, 1, 12),
        "dow": _parse_field(dow, 0, 6),
    }


def _matches(dt, spec):
    if dt.minute not in spec["minute"] or dt.hour not in spec["hour"] or dt.month not in spec["month"]:
        return False
    dom_ok = dt.day in spec["dom"]
    dow_ok = (dt.weekday() + 1) % 7 in spec["dow"]
    if spec["dom"] != set(range(1, 32)) and spec["dow"] != set(range(0, 7)):
        return dom_ok or dow_ok
    return dom_ok and dow_ok


def _next_run_after(cron_expression, after, limit_minutes=527040):
    """Smallest minute-aligned datetime strictly after `after` matching the cron spec."""
    spec = _parse_cron(cron_expression)
    cursor = after.replace(second=0, microsecond=0) + timedelta(minutes=1)
    for _ in range(limit_minutes):
        if _matches(cursor, spec):
            return cursor
        cursor += timedelta(minutes=1)
    raise RuntimeError("No matching run found within search window")


def expected_interval_ms(cron_expression, anchor):
    """The gap, in ms, between two consecutive matches of cron_expression near anchor."""
    first_run = _next_run_after(cron_expression, anchor)
    second_run = _next_run_after(cron_expression, first_run)
    return (second_run - first_run).total_seconds() * 1000


def is_scheduler_stalled(last_run_at, now, cron_schedule, tolerance_multiplier=3):
    """Pure: no I/O. Returns True iff the gap since last_run_at exceeds the
    schedule's expected interval times tolerance_multiplier."""
    interval_ms = expected_interval_ms(cron_schedule, last_run_at)
    gap_ms = (now - last_run_at).total_seconds() * 1000
    return gap_ms > interval_ms * tolerance_multiplier


def alert_stalled_scheduler(gap_minutes, webhook_url=None):
    message = (
        f"Medusa scheduler looks stalled. No heartbeat for {gap_minutes:.1f} minutes. "
        "A workflow step is likely stuck occupying the only BullMQ worker slot "
        "(jobWorkerOptions.concurrency=1). Restart the worker process "
        "(MEDUSA_WORKER_MODE=worker) to recover. Consider raising concurrency and "
        "adding step-level timeouts to prevent this recurring."
    )
    if webhook_url:
        requests.post(webhook_url, json={"text": message}, timeout=15)
    return message


def run():
    token = get_token()
    last_run_iso = get_heartbeat_last_run_at(token, STOCK_LOCATION_ID)
    now = datetime.now(timezone.utc)

    if last_run_iso is None:
        log.warning("No heartbeat recorded yet at all. Treating as stalled.")
        message = alert_stalled_scheduler(float("inf"), ALERT_WEBHOOK_URL if not DRY_RUN else None)
        log.warning(message)
        return

    last_run_at = datetime.fromisoformat(last_run_iso.replace("Z", "+00:00"))
    stalled = is_scheduler_stalled(last_run_at, now, HEARTBEAT_CRON, TOLERANCE_MULTIPLIER)

    gap_minutes = (now - last_run_at).total_seconds() / 60
    if not stalled:
        log.info("Scheduler healthy. Last heartbeat %.1f minute(s) ago.", gap_minutes)
        return

    log.warning("Scheduler stalled. Last heartbeat %.1f minute(s) ago.", gap_minutes)
    message = alert_stalled_scheduler(gap_minutes, ALERT_WEBHOOK_URL if not DRY_RUN else None)
    log.warning(message)


if __name__ == "__main__":
    run()

File: scheduled-jobs-stop-after-uptime/node/check-scheduler-heartbeat.js
/**
 * Detect a stalled Medusa v2 scheduler caused by a hung workflow step
 * occupying the only BullMQ worker slot (jobWorkerOptions.concurrency=1
 * by default on @medusajs/medusa/workflow-engine-redis). There is no
 * Admin API route that can kill a stuck job or restart the scheduler,
 * so this only flags the stall and alerts an operator to restart the
 * worker process. DRY_RUN=true only logs locally; DRY_RUN=false also
 * calls the alert webhook. Never writes anything to Medusa itself.
 */
import { pathToFileURL } from "node:url";

const BASE_URL = process.env.MEDUSA_BACKEND_URL || "http://localhost:9000";
const EMAIL = process.env.MEDUSA_ADMIN_EMAIL || "admin@example.com";
const PASSWORD = process.env.MEDUSA_ADMIN_PASSWORD || "supersecret";
const STOCK_LOCATION_ID = process.env.HEARTBEAT_STOCK_LOCATION_ID || "sloc_heartbeat";
const HEARTBEAT_CRON = process.env.HEARTBEAT_CRON || "*/5 * * * *";
const TOLERANCE_MULTIPLIER = Number(process.env.TOLERANCE_MULTIPLIER || 3);
const ALERT_WEBHOOK_URL = process.env.ALERT_WEBHOOK_URL || "";
const DRY_RUN = (process.env.DRY_RUN || "true").toLowerCase() === "true";

function parseField(field, lo, hi) {
  if (field === "*") return new Set(Array.from({ length: hi - lo + 1 }, (_, i) => lo + i));
  const values = new Set();
  for (const part of field.split(",")) {
    if (part.startsWith("*/")) {
      const step = Number(part.slice(2));
      for (let v = lo; v <= hi; v += step) values.add(v);
    } else if (part.includes("-")) {
      const [a, b] = part.split("-").map(Number);
      for (let v = a; v <= b; v++) values.add(v);
    } else {
      values.add(Number(part));
    }
  }
  return values;
}

function parseCron(cronExpression) {
  const [minute, hour, dom, month, dow] = cronExpression.trim().split(/\s+/);
  return {
    minute: parseField(minute, 0, 59),
    hour: parseField(hour, 0, 23),
    dom: parseField(dom, 1, 31),
    month: parseField(month, 1, 12),
    dow: parseField(dow, 0, 6),
  };
}

function fullRange(lo, hi) {
  return new Set(Array.from({ length: hi - lo + 1 }, (_, i) => lo + i));
}

function matches(date, spec) {
  if (!spec.minute.has(date.getUTCMinutes())) return false;
  if (!spec.hour.has(date.getUTCHours())) return false;
  if (!spec.month.has(date.getUTCMonth() + 1)) return false;
  const domOk = spec.dom.has(date.getUTCDate());
  const dowOk = spec.dow.has(date.getUTCDay());
  const domIsFull = spec.dom.size === fullRange(1, 31).size;
  const dowIsFull = spec.dow.size === fullRange(0, 6).size;
  if (!domIsFull && !dowIsFull) return domOk || dowOk;
  return domOk && dowOk;
}

function nextRunAfter(cronExpression, after, limitMinutes = 527040) {
  // Smallest minute-aligned date strictly after `after` matching the cron spec.
  const spec = parseCron(cronExpression);
  const cursor = new Date(after.getTime());
  cursor.setUTCSeconds(0, 0);
  cursor.setUTCMinutes(cursor.getUTCMinutes() + 1);
  for (let i = 0; i < limitMinutes; i++) {
    if (matches(cursor, spec)) return new Date(cursor.getTime());
    cursor.setUTCMinutes(cursor.getUTCMinutes() + 1);
  }
  throw new Error("No matching run found within search window");
}

export function expectedIntervalMs(cronExpression, anchor) {
  // The gap, in ms, between two consecutive matches of cronExpression near anchor.
  const firstRun = nextRunAfter(cronExpression, anchor);
  const secondRun = nextRunAfter(cronExpression, firstRun);
  return secondRun.getTime() - firstRun.getTime();
}

export function isSchedulerStalled(lastRunAt, now, cronSchedule, toleranceMultiplier = 3) {
  // Pure: no I/O. True iff the gap since lastRunAt exceeds the schedule's
  // expected interval times toleranceMultiplier.
  const intervalMs = expectedIntervalMs(cronSchedule, lastRunAt);
  const gapMs = now.getTime() - lastRunAt.getTime();
  return gapMs > intervalMs * toleranceMultiplier;
}

async function getToken() {
  const res = await fetch(`${BASE_URL}/auth/user/emailpass`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ email: EMAIL, password: PASSWORD }),
  });
  if (!res.ok) throw new Error(`Medusa auth ${res.status}`);
  const body = await res.json();
  return body.token;
}

async function getHeartbeatLastRunAt(token, stockLocationId) {
  const res = await fetch(
    `${BASE_URL}/admin/stock-locations?fields=id,metadata&limit=100`,
    { headers: { Authorization: `Bearer ${token}` } }
  );
  if (!res.ok) throw new Error(`Medusa ${res.status}`);
  const body = await res.json();
  const loc = body.stock_locations.find((l) => l.id === stockLocationId);
  return loc?.metadata?.heartbeat_last_run_at ?? null;
}

async function alertStalledScheduler(gapMinutes, webhookUrl) {
  const message =
    `Medusa scheduler looks stalled. No heartbeat for ${gapMinutes.toFixed(1)} minutes. ` +
    "A workflow step is likely stuck occupying the only BullMQ worker slot " +
    "(jobWorkerOptions.concurrency=1). Restart the worker process " +
    "(MEDUSA_WORKER_MODE=worker) to recover. Consider raising concurrency and " +
    "adding step-level timeouts to prevent this recurring.";
  if (webhookUrl) {
    await fetch(webhookUrl, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ text: message }),
    });
  }
  return message;
}

export async function run() {
  const token = await getToken();
  const lastRunIso = await getHeartbeatLastRunAt(token, STOCK_LOCATION_ID);
  const now = new Date();

  if (!lastRunIso) {
    console.warn("No heartbeat recorded yet at all. Treating as stalled.");
    const message = await alertStalledScheduler(Infinity, DRY_RUN ? "" : ALERT_WEBHOOK_URL);
    console.warn(message);
    return;
  }

  const lastRunAt = new Date(lastRunIso);
  const stalled = isSchedulerStalled(lastRunAt, now, HEARTBEAT_CRON, TOLERANCE_MULTIPLIER);
  const gapMinutes = (now.getTime() - lastRunAt.getTime()) / 60000;

  if (!stalled) {
    console.log(`Scheduler healthy. Last heartbeat ${gapMinutes.toFixed(1)} minute(s) ago.`);
    return;
  }

  console.warn(`Scheduler stalled. Last heartbeat ${gapMinutes.toFixed(1)} minute(s) ago.`);
  const message = await alertStalledScheduler(gapMinutes, DRY_RUN ? "" : ALERT_WEBHOOK_URL);
  console.warn(message);
}

if (import.meta.url === pathToFileURL(process.argv[1]).href) {
  run().catch((err) => { console.error(err); process.exit(1); });
}
